### PR TITLE
(plugin) cloud::azure::managemement::budgets - Add cost management plugin

### PR DESCRIPTION
**Pull request by matoy**
## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [X] 21.04.x
- [X] 21.10.x
- [X] 22.04.x
- [X] 22.10.x (master)

<h2> How this pull request can be tested ? </h2>

This plugin allows to query the actual cost for a subscription/resource group and compare it to a given Azure budget.

/usr/lib/centreon/plugins/centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::management::costs::plugin --custommode='api' --mode='cost-usage' --subscription='xxxx' --tenant='xxxx' --client-id='xxxx' --client-secret='xxxx' --budget-name='MyMonthlyBudget' --warning-cost='95' --critical-cost='99'
OK: Spent amount is 1400.25EUR on 1500EUR of allowed budget (93.33% consumption) for the past 30 days | 'azure.budget.cost'=1400.25;0:1425;0:1485;0;1500

/usr/lib/centreon/plugins/centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::management::costs::plugin --custommode='api' --mode='cost-usage' --subscription='xxxx' --tenant='xxxx' --client-id='xxxx' --client-secret='xxxx' --budget-name='MyMonthlyBudget' --warning-cost='90' --critical-cost='99'
WARNING: Spent amount is 1400.25EUR on 1500EUR of allowed budget (93.33% consumption) for the past 30 days | 'azure.budget.cost'=1400.25;0:1350;0:1485;0;1500

#### Community contributors & Centreon team

- [X] I have followed the **coding style guidelines** provided by Centreon
- [X] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [X] I have commented my code, especially **hard-to-understand areas** of the PR.
- [X] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by Sims24** [comment]
Hi @matoy 

Once again, thanks fot taking time to share your work with the community. 

I'll put some inline comments, let me know your thoughts about it. Then I'll push on your branch some cosmetic change (mode name, indentation, etc.). 

Thanks a lot!

File: cloud/azure/management/costs/mode/costusage.pm [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    my ($self, %options) = @_;
+
+    my %budget_options = ();
+    if ($self->{instance_mode}->{option_results}->{units} eq '%') {
+        $budget_options{total} = $self->{result_values}->{budget};
+        $budget_options{cast_int} = 1;
+    }
+
+    $self->{output}->perfdata_add(
+        nlabel => 'azure.budget.cost',
```
Comment (Sims24) [inline cloud/azure/management/costs/mode/costusage.pm:38]: azure.budget.consumption.currency

File: cloud/azure/management/costs/mode/costusage.pm [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    if (!$budget) {
+	$self->{output}->add_option_msg(short_msg => "No " . $self->{option_results}->{budget_name} . " found (or missing permissions)");
+	$self->{output}->option_exit();
+    }
+    if (!$costs || $cost < 0.01) {
+	$self->{output}->add_option_msg(short_msg => "Null or < 0.01 " . $budget->{properties}->{currentSpend}->{unit} . "cost found on the specified scope.");
+	$self->{output}->option_exit();
+    }
```
Comment (Sims24) [inline cloud/azure/management/costs/mode/costusage.pm:160]: @matoy , would it be a problem to force à 0 here? Could it hide some errors (mismatch resourcegroup vs budget)?

File: cloud/azure/management/costs/plugin.pm [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+use base qw(centreon::plugins::script_custom);
+
+sub new {
+    my ( $class, %options ) = @_;
+    my $self = $class->SUPER::new( package => __PACKAGE__, %options );
+    bless $self, $class;
+
+    $self->{version} = '0.1';
+    %{ $self->{modes} } = (
+        'cost-usage'            => 'cloud::azure::management::costs::mode::costusage'
```
Comment (Sims24) [inline cloud/azure/management/costs/plugin.pm:34]: we don't add 'usage' anymore. I think a will simply rename it to budgets.

**Review by Sims24 — COMMENTED** [review]
Ref inline comments & questions.

File: cloud/azure/custom/api.pm [hunk 4]
```diff
@@ -790,6 +790,56 @@ sub azure_get_publicip {
     return $response;
 }
 
+sub azure_get_budget_set_url {
+    my ($self, %options) = @_;
+
+    my $url = $self->{management_endpoint} . "/subscriptions/" . $self->{subscription};
+    $url .= "/resourceGroups/" . $options{resource_group} if (defined($options{resource_group}) && $options{resource_group} ne '');
+    $url .= "/providers/Microsoft.Consumption/budgets/" . $options{budget_name} if (defined($options{budget_name}) && $options{budget_name} ne '');
```
Comment (Sims24) [inline cloud/azure/custom/api.pm:798]: Budget name is already mandatory at the mode level, I think we don't need another if defined(budget) check right?

File: cloud/azure/custom/api.pm [hunk 5]
```diff
@@ -790,6 +790,56 @@ sub azure_get_publicip {
...
+
+    return $response;
+}
+
+sub azure_get_usagedetails_set_url {
+    my ($self, %options) = @_;
+
+    my $url = $self->{management_endpoint} . "/subscriptions/" . $self->{subscription};
+    $url .= "/resourceGroups/" . $options{resource_group} if (defined($options{resource_group}) && $options{resource_group} ne '');
+    $url .= "/providers/Microsoft.Consumption/usageDetails?\$filter=properties%2FusageStart ge %27" . $options{usage_start} . "%27 and properties%2FusageEnd le %27" . $options{usage_end} . "%27" if (defined($options{usage_start}) && $options{usage_start} ne '' && defined($options{usage_end}) && $options{usage_end} ne '');
```
Comment (Sims24) [inline cloud/azure/custom/api.pm:818]: Same comment here nope: https://github.com/centreon/centreon-plugins/pull/3866/files#r959273964  ?

**Comment by Sims24** [comment]
One last (tiny) thing, could you rename the costs mode to budgets, as it will better relates to the API and probe parameters. 

Thanks again @matoy

**Comment by Sims24** [comment]
Finally did it. 

Thanks!

**Review by Sims24 — APPROVED** [review]

**Comment by matoy** [comment]
Hi,

There's still:
  %{ $self->{modes} } = (
        '**costs**' => 'cloud::azure::management::costs::mode::budgets'
    );

i guess you want to rename it here too

**Pull request merged**